raven_aiohttp.py

```diff
@@ -58,50 +80,129 @@ def keepalive(self):
     def family(self):
         return self._family
 
-    def client_session(self):
-        if not self.keepalive or \
-                (self._client_session is None or self._client_session.closed):
-            connector = aiohttp.TCPConnector(verify_ssl=self.verify_ssl,
-                                             resolve=self.resolve,
-                                             family=self.family,
-                                             loop=self._loop)
-            self._client_session = aiohttp.ClientSession(connector=connector,
-                                                         loop=self._loop)
-        return self._client_session
+    def _client_session_factory(self):
+        # by default aiohttp.TCPConnector deadly caches dns
+        connector = aiohttp.TCPConnector(verify_ssl=self.verify_ssl,
+                                         resolve=self.resolve,
+                                         family=self.family,
+                                         use_dns_cache=False,
+                                         loop=self._loop)
+        return aiohttp.ClientSession(connector=connector,
+                                     loop=self._loop)
+
+    @asyncio.coroutine
+    def close(self, timeout=None, return_exceptions=True):
+        self._closing = True
+
+        try:
+            with async_timeout.timeout(timeout, loop=self._loop):
+                if self._background_workers:
+                    yield from self._queue.put(...)
+
+                    yield from asyncio.gather(
+                        *self._workers,
+                        return_exceptions=return_exceptions,
+                        loop=self._loop
+                    )
+
+                    assert len(self._workers) == 0
+                else:
+                    yield from asyncio.gather(
+                        *self._tasks,
+                        return_exceptions=return_exceptions,
+                        loop=self._loop
+                    )
+
+                    assert len(self._tasks) == 0
+        except asyncio.TimeoutError:
+            pass
+        finally:
+            if self.keepalive:
```

Comment:
Cancel all not finished yet tasks/workers?

Reply:
Tasks not cancelled, we are waiting for them to be finished, a well same for workers.
If You want to stop it immediately `timeout=0` should be used

Reply:
Ok

Reply:
What about to make default timeout '0' ?

Reply:
Not sure.
After stopping the server sending all pending events to sentry server makes sense.
It should be pretty fast procedure.